### ISSUE TITLE
the big different of outputs between the .pth model and the according onnx model with the BatchNorm2d layer

### ISSUE THREAD

**Issue by wuqingshan2010** [microsoft/onnxruntime#3873]
**Describe the bug**
when i export the .pth model to .onnx model, and load the .onnx model with the onnxruntime,  i get the big different of outputs between them. the .pth model contains the BatchNorm2d layer.  when i remove the BatchNorm2d layer, the different of outputs is smaller.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):window 10, x64
- ONNX Runtime installed from (source or binary): binary, pip install
- ONNX Runtime version:1.2.0
- Python version:3.7

def conv_bn(inp, oup, stride):
    return nn.Sequential(
        nn.Conv2d(inp, oup, 3, stride, 1, bias=False),
        nn.BatchNorm2d(oup),
        nn.ReLU6(inplace=True)
    )

**Comment by hariharans29** [comment]
Can you share both the models ?

**Comment by hariharans29** [comment]
Hi @wuqingshan2010 ,

Any update on this ? Could you please share both the models  (one with BatchNorm and one without) ? This seems to be some sort of exporter issue but still would like to make sure it is.

**Comment by OronG13** [comment]
Hello,
I have the same problem.

Attached is a script that I wrote which demonstrate the problem.
[TensorRTOnnxAccuracyTests.txt](https://github.com/microsoft/onnxruntime/files/5382375/TensorRTOnnxAccuracyTests.txt)

 I'm working with PyTorch 1.1.0 and OnnxRuntime 1.5.1 on Windows.

Any help here to find the problem root cause or other way to achieve an onnx model which provide the same results as the PyTorch model.

Thanks,

**Comment by hariharans29** [comment]
Hello all,

Thanks for the repro script @OronG13.

I investigated this and this is why I think it is an exporter issue to be raised with Torch:

I modified the repro script provided above so that the model only contains 1 BatchNorm node with input dimensions that may be manually managed quite easily: [1, 1, 2, 2] : [N, C, H, W]

The random input to the BatchNorm node was: [[[[2.2950,  -0.0380], [1.3294,  -1.0175]]]], shape : [1, 1, 2, 2] (N, C, H, W).

The manually computed mean across the single channel (across H and W) is: 0.642225

The manually computed variance across the single channel (across H and W) is: 1.605325

This is what I would have expected to have seen in the exported ONNX model as well. Yet the exported ONNX model has this:

![image](https://user-images.githubusercontent.com/9969784/97056265-6bc20d80-153d-11eb-8605-8c12da62fc41.png)

Both mean and variance are off compared to the manually computed values.

When I override the model's mean and var with my manually computed mean and var (above), the manually computed results (and ORT’s results) align with Torch’s results exactly. So, we need to understand the exporter’s rationale for the mean and variance it has stamped to this node.

**Comment by OronG13** [comment]
Thank you **@hariharans29** ,
Your experiment results are very interesting and make a significant step forward in this entire investigation on the BacthNorm2D differences between PyTorch and ORT.

I just want to ask please, when you say "...we need to understand...":
Can you clarify what are the exactly steps that shall be done?
Who will take the responsibility to do them?
Can I help someway?

Regards,

**Comment by hariharans29** [comment]
I shared the same data with the PyTorch ONNX exporter group. CC: @spandantiwari

**Comment by OronG13** [comment]
Hello,
It will be much appropriated if you will update with the investigation process status please.
Meantime, as you recommended I opened an issue in the PyTorch forum which reference here.
[https://discuss.pytorch.org/t/inference-model-with-batchnorm2d-generates-different-output-via-pytorch-and-onnxruntime/101316](url)
Thanks,

**Comment by hariharans29** [comment]
Hello,

I am afraid, at this point, I am not sure how to proceed with the investigation. For the parameters provided in the exported ONNX model, ORT seems to crunch the numbers correctly (i.e.) there doesn’t seem to be any bug in ORT that is causing the data mismatch. As to why Torch’s parameters are different while inferencing its model from the exported ONNX model’s parameters, it is beyond my scope of knowledge. Sorry.

**Comment by hariharans29** [comment]
Can you please reference this issue in the issue you opened against PyTorch ? That way, their devs might be able to be able to instantly recognize what’s wrong based on the shared data.

**Comment by OronG13** [comment]
Hi **@hariharans29**
Thanks for your response.

I thought that you may have an access to the PyTorch ONNX exporter group...do you?
I added to the PyTorch opened issue a reference to this issue here...

**Comment by hariharans29** [comment]
Hi @OronG13 - Sorry I didn't see your response. I shared my findings with them (CC'ed @spandantiwari in the previous message). Did you open a PyTorch issue and reference this ? (I don't see any issue referencing this - hence checking)

**Comment by OronG13** [comment]
Hello @hariharans29,

Thank you for continuing monitor this important issue.

Yes, I opened an issue in the PyTorch forum (15 days ago);
[https://discuss.pytorch.org/t/inference-model-with-batchnorm2d-generates-different-output-via-pytorch-and-onnxruntime/101316](url)

I didn't get any response till now...

This issue is one of our major obstacle and it is very imporatant for us continuing with our CNN model development program.
Due to the fact that it includes several BatchNorm2D layers we are totally depends on a solution for this problem.

Thanks,

**Comment by deepeshguru** [comment]
I am also seen same issue, output is mismatch for BatchNorm1d, BatchNorm2d and BatchNorm3d. I am passing same input in pytorch model and onnx model. 
I am using pytorch version:  1.7.0 and onnxruntime version:  1.5.2

**code snippet for convert pytorch batchnorm1d to onnx:**
import torch
import numpy as np
np.random.seed(0)
example_input = np.random.randn(3, 3).astype(np.float32)
example_input = torch.from_numpy(example_input)
batchnorm1d = torch.nn.BatchNorm1d(3)
batchnorm1d_output_pytorch = batchnorm1d(example_input)
torch.onnx.export(batchnorm1d, example_input, "./models/BatchNorm1d.onnx", verbose=False)
print("pytorch batchnorm1d input: ", example_input)
print("pytorch batchnorm1d output: ", batchnorm1d_output_pytorch)
**output of above code:**
pytorch batchnorm1d input:  tensor([[ 1.7641,  0.4002,  0.9787],
        [ 2.2409,  1.8676, -0.9773],
        [ 0.9501, -0.1514, -0.1032]])
pytorch batchnorm1d output:  tensor([[ 0.2109, -0.3583,  1.2657],
        [ 1.1056,  1.3639, -1.1791],
        [-1.3165, -1.0056, -0.0866]], grad_fn=NativeBatchNormBackward)

**code snippet for loading batchnorm1d onnx:**
import onnxruntime as ort
import numpy as np
np.random.seed(0)
example_input = np.random.randn(3, 3).astype(np.float32)
ort_session = ort.InferenceSession('./models/BatchNorm1d.onnx')
batchnorm1d_output_onnx = ort_session.run(None, {'input': example_input})
print("onnxrt batchnorm1d input: ", example_input)
print("onnxrt batchnorm1d output: ", batchnorm1d_output_onnx)
**output of above code:**
onnxrt batchnorm1d input:  [[ 1.7640524   0.4001572   0.978738  ]
 [ 2.2408931   1.867558   -0.9772779 ]
 [ 0.95008844 -0.1513572  -0.10321885]]
onnxrt batchnorm1d output:  [array([[ 1.6468376 ,  0.32815474,  0.9840902 ],
       [ 2.1379795 ,  1.7890683 , -0.9758297 ],
       [ 0.8084616 , -0.22092146, -0.10002615]], dtype=float32)]

**code snippet for calculating difference between both models output:**
diff = np.sum(np.absolute(batchnorm1d_output_onnx - batchnorm1d_output_pytorch.detach().numpy()))
print(diff)
**output of above code:**
6.9879394

**Comment by smaximz** [comment]
Hello, is there any progress in this issue?
PyTorch issue didn't get any response.
Maybe you can help to promote this issue in one of the websites @spandantiwari , @hariharans29

**Comment by OronG13** [comment]
Hello,
Any kind of support will be much appreciated.

It is a major concern that we cannot include BatchNorm2D operator in our models due to *.onnx export problem.

In order to be able using NVIDIA TensorRT sdk, we shall understand what is the root cause of the wrong BatchNorm2D behavior after converting the TorchScript *.pt to *.onnx format.

Thank,

**Comment by hariharans29** [comment]
I raised this issue again. Sorry for the delay in response.

**Comment by KsenijaS** [comment]
@OronG13 can you please add model.eval() before running pytorch model on line 35. That should fix the error that you are getting. Thanks!

**Comment by OronG13** [comment]
Hello @KsenijaS,
It works!
Thank you all very much for your support.

I will take this change to my full model Env. and test it if it works well also and update you.

Regard,

**Comment by hariharans29** [comment]
Thanks a lot for taking a look @KsenijaS !

**Comment by MInner** [comment]
For future record,  I can confirm that the @deepeshguru's snipped above produces identical results **with `model.eval()`**

```
import torch
import numpy as np
np.random.seed(0)
example_input = np.random.randn(3, 3).astype(np.float32)
example_input = torch.from_numpy(example_input)
batchnorm1d = torch.nn.BatchNorm1d(3)
batchnorm1d.eval()
batchnorm1d_output_pytorch = batchnorm1d(example_input)
torch.onnx.export(batchnorm1d, example_input, "./BatchNorm1d.onnx", verbose=False)
print("pytorch batchnorm1d input: ", example_input)
print("pytorch batchnorm1d output: ", batchnorm1d_output_pytorch)

import onnxruntime as ort
import numpy as np
np.random.seed(0)
example_input = np.random.randn(3, 3).astype(np.float32)
ort_session = ort.InferenceSession('./BatchNorm1d.onnx')
batchnorm1d_output_onnx = ort_session.run(None, {'input': example_input})
print("onnxrt batchnorm1d input: ", example_input)
print("onnxrt batchnorm1d output: ", batchnorm1d_output_onnx)
```

```
pytorch batchnorm1d input:  tensor([[ 1.7641,  0.4002,  0.9787],
        [ 2.2409,  1.8676, -0.9773],
        [ 0.9501, -0.1514, -0.1032]])
pytorch batchnorm1d output:  tensor([[ 1.7640,  0.4002,  0.9787],
        [ 2.2409,  1.8675, -0.9773],
        [ 0.9501, -0.1514, -0.1032]], grad_fn=<NativeBatchNormBackward0>)
onnxrt batchnorm1d input:  [[ 1.7640524   0.4001572   0.978738  ]
 [ 2.2408931   1.867558   -0.9772779 ]
 [ 0.95008844 -0.1513572  -0.10321885]]
onnxrt batchnorm1d output:  [array([[ 1.7640436 ,  0.40015522,  0.9787331 ],
       [ 2.240882  ,  1.8675487 , -0.977273  ],
       [ 0.9500837 , -0.15135644, -0.10321834]], dtype=float32)]
```
